crates/core/tedge_watchdog/src/systemd_watchdog.rs

```diff
@@ -233,47 +231,37 @@ async fn monitor_tedge_service(
         }
 
         let elapsed = start.elapsed();
-        if elapsed < tokio::time::Duration::from_secs(interval) {
-            tokio::time::sleep(tokio::time::Duration::from_secs(interval) - elapsed).await;
+        if elapsed < Duration::from_secs(interval) {
+            tokio::time::sleep(Duration::from_secs(interval) - elapsed).await;
             warn!("tedge systemd watchdog not started because no services to monitor")
         }
     }
 }
 
 async fn get_latest_health_status_message(
-    request_timestamp: String,
+    request_timestamp: OffsetDateTime,
     messages: &mut mpsc::UnboundedReceiver<Message>,
 ) -> Result<HealthStatus, WatchdogError> {
-    loop {
-        if let Some(message) = messages.next().await {
-            if let Ok(message) = message.payload_str() {
-                debug!("Health response received: {}", message);
-                if let Ok(health_status) = serde_json::from_str::<HealthStatus>(message) {
-                    let request_timestamp = OffsetDateTime::parse(
-                        &request_timestamp,
-                        &format_description::well_known::Rfc3339,
-                    )?
-                    .unix_timestamp();
-                    let datetime = OffsetDateTime::parse(
-                        &health_status.time,
-                        &format_description::well_known::Rfc3339,
-                    )?
-                    .unix_timestamp();
-
-                    if datetime >= request_timestamp {
-                        return Ok(health_status);
-                    } else {
-                        debug!(
-                            "Ignoring stale health response: {:?} older than request time: {}",
-                            health_status, request_timestamp
-                        );
-                    }
+    while let Some(message) = messages.next().await {
+        if let Ok(message) = message.payload_str() {
+            debug!("Health response received: {message}");
+            if let Ok(health_status) = serde_json::from_str::<HealthStatus>(message) {
+                let datetime = IsoOrUnix::try_from(&health_status.time)?;
+
+                // Compare to a slightly old timestamp to avoid false negatives from floating-point error in unix timestamps
+                if datetime.into_inner() >= request_timestamp - Duration::from_millis(10) {
```

Comment:
>  I am assuming this is only comparing timestamps from the same machine

This is a fair assumption, as the `systemd_watchdog` can only watch processes running on the same box.

> More generally, this doesn't feel like an ideal solution to the problem since the system time is not monotonic and could change between the healthcheck request being sent and the reply being received, although we weren't previously robust to this.

Yeah, an option could be to use `/proc/uptime` for this check. But this is beyond the scope of this PR.

I'm okay with the current proposal.